notebooks/108-gpu-device/108-gpu-device.ipynb

```diff
@@ -0,0 +1,1009 @@
+    "    \"toaster\", \"sink\", \"refrigerator\", \"blender\", \"book\", \"clock\", \"vase\", \"scissors\",\n",
+    "    \"teddy bear\", \"hair drier\", \"toothbrush\", \"hair brush\"\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0ad10b97-e32e-4885-af1d-0636e970b707",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "### Set up Asynchronous Pipeline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "91f6ed58-8b3f-415e-a733-a7517e200b12",
+   "metadata": {},
+   "source": [
+    "#### Callback Definition"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1268e1a4-b530-4b39-b79d-d6d5f988a077",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define a callback function that runs every time the asynchronous pipeline completes inference on a frame\n",
+    "def completion_callback(infer_request: InferRequest, frame_id: int) -> None:\n",
+    "    global frame_number\n",
+    "    stop_time = time.time()\n",
+    "    frame_number += 1\n",
+    "\n",
+    "    predictions = next(iter(infer_request.results.values()))\n",
+    "    results[frame_id] = predictions[:10] # Grab first 10 predictions for this frame\n",
+    "    \n",
+    "    total_time = stop_time - start_time\n",
+    "    frame_fps[frame_id] = frame_number / total_time"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2d2f6aba-b3ba-478b-b02f-6b438b575c33",
+   "metadata": {},
+   "source": [
+    "#### Create Async Pipeline"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "56e8eae2-20a9-43dc-8774-444ad0179136",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create asynchronous inference queue with optimal number of infer requests\n",
+    "infer_queue = AsyncInferQueue(compiled_model)\n",
+    "infer_queue.set_callback(completion_callback)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fb47ae59-b86d-46c4-b219-e7beccd5f796",
+   "metadata": {},
+   "source": [
+    "### Perform Inference"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "59f493dd-4cb7-4ae2-b117-7ba41a950d86",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Perform inference on every frame in the framebuffer\n",
+    "results = {}\n",
+    "frame_fps = {}\n",
+    "frame_number = 0\n",
+    "start_time = time.time()\n",
+    "for i, input_frame in enumerate(framebuf):\n",
+    "    infer_queue.start_async({0: input_frame}, i)\n",
+    "\n",
+    "infer_queue.wait_all() # Wait until all inference requests in the AsyncInferQueue are completed\n",
+    "stop_time = time.time()\n",
+    "\n",
+    "# Calculate total inference time and FPS\n",
+    "total_time = stop_time - start_time\n",
+    "fps = len(framebuf) / total_time\n",
+    "time_per_frame = 1 / fps \n",
+    "print(f'Total time to infer all frames: {total_time:.3f}s')\n",
+    "print(f'Time per frame: {time_per_frame:.6f}s ({fps:.3f} FPS)')"
```

Comment:
Hi actually the ```time per frame``` here means the throughput, how about latency ? shall we also record the average inference time for each frame ?

Reply:
Let's do it as follow up